contracts/PushStaking/PushStaking.sol

```diff
@@ -0,0 +1,549 @@
+        returns (uint256 sharesToBeAllocated)
+    {
+        if (_percentage.percentageNumber / 10 ** _percentage.decimalPlaces >= 100 || _percentage.percentageNumber == 0) {
+            revert Errors.InvalidArg_MoreThanExpected(99, _percentage.percentageNumber);
+        }
+        sharesToBeAllocated = (_percentage.percentageNumber * _totalShares)
+            / ((100 * (10 ** _percentage.decimalPlaces)) - _percentage.percentageNumber);
+    }
+
+    /*
+     * Adding Wallet Share to a Wallet
+     * 1. addWalletShare(address wallet, uint256 percentageOfShares)
+     * 2. Can be called by governance.
+     * 3. Uses the formulae to derive the percent of shares to be assigned to a specific wallet
+     * 4. Updates WALLET_TOTAL_SHARES
+     * 5. Updates walletShareInfo mapping
+     * 6. Emits out an event.
+    * 7. If a wallet has already has a share, then it acts as a "increase share" function. And the percenatge passed
+     *    should be greater than the already assigned percentge.
+    */
+    function addWalletShare(address _walletAddress, GenericTypes.Percentage memory _percentage) public onlyGovernance {
+        if(_walletAddress == address(0)){
+            revert Errors.InvalidArgument_WrongAddress(_walletAddress);
+        }
+        uint256 TotalShare = WALLET_TOTAL_SHARES;
+        uint256 currentWalletShare = walletShareInfo[_walletAddress].walletShare;
+        if (currentWalletShare != 0) {
+            TotalShare -= currentWalletShare;
+        }
+        uint256 sharesToBeAllocated = getSharesAmount(TotalShare, _percentage);
+        if (sharesToBeAllocated <= currentWalletShare) {
+            revert Errors.InvalidArg_LessThanExpected(currentWalletShare, sharesToBeAllocated);
+        }
+        walletShareInfo[_walletAddress].lastClaimedBlock = walletShareInfo[_walletAddress].lastClaimedBlock == 0
+            ? genesisEpoch
+            : walletShareInfo[_walletAddress].lastClaimedBlock;
+        _adjustWalletAndTotalStake(_walletAddress, sharesToBeAllocated, currentWalletShare);
+        WALLET_TOTAL_SHARES = TotalShare + sharesToBeAllocated;
+        emit NewSharesIssued(_walletAddress, sharesToBeAllocated);
+    }
+    /*
+     * Removing Wallet Share from a Wallet
+     * 1. removes the shares from a wallet completely
+     * 2. Can be called by governance.
+     * 3. Updates WALLET_TOTAL_SHARES
+     * 4. Emits out an event.
+    */
+
+    function removeWalletShare(address _walletAddress) public onlyGovernance {
+        if(_walletAddress == address(0) || _walletAddress == FOUNDATION) {
+            revert Errors.InvalidArgument_WrongAddress(_walletAddress);
+        }
+        uint256 sharesToBeRemoved = walletShareInfo[_walletAddress].walletShare;
+        _adjustWalletAndTotalStake(_walletAddress, 0, sharesToBeRemoved);
+        _adjustWalletAndTotalStake(FOUNDATION, sharesToBeRemoved, 0);
+
+        emit SharesRemoved(_walletAddress, sharesToBeRemoved);
+    }
+
+    function decreaseWalletShare(
```

Comment:
Important:
- Decrement of Wallet Share should follow the procedure mentioned in this doc - https://www.notion.so/pushprotocol/On-Removal-Decrement-of-Wallet-Shares-12f188aea7f480b0b505f8c6d506b347
- Ideally all relinquished shares should just go to foundation and no change in share percentage should occur in existing wallets. @0xNilesh @Zartaj0